File: app/src/main/java/org/miaowo/miaowo/adapter/ChatPagerAdapter.java
package org.miaowo.miaowo.adapter;

import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentPagerAdapter;

import org.miaowo.miaowo.bean.data.ChatMessage;
import org.miaowo.miaowo.bean.data.ChatRoom;
import org.miaowo.miaowo.bean.data.User;
import org.miaowo.miaowo.fragment.chat.ChatFragment;
import org.miaowo.miaowo.fragment.chat.ChatListFragment;

import java.util.ArrayList;
import java.util.List;

/**
 * ViewPager 的 Adapter
 * Created by luqin on 17-4-23.
 */

public class ChatPagerAdapter extends FragmentPagerAdapter {
    private List<ChatFragment> mFragments;
    private ChatListFragment mListFragment;

    public ChatPagerAdapter(FragmentManager fm) {
        super(fm);
        mFragments = new ArrayList<>();
        mListFragment = ChatListFragment.newInstance();
    }

    @Override
    public Fragment getItem(int position) {
        if (position == 0) return mListFragment;
        else return mFragments.get(position - 1);
    }

    @Override
    public int getCount() {
        return mFragments.size() + 1;
    }

    public int addChat(ChatRoom room) {
        ChatFragment fragment = ChatFragment.newInstance(room);
        mFragments.add(fragment);
        return mFragments.indexOf(fragment) + 1;
    }

    public int getPosition(int roomId) {
        for (ChatFragment fragment : mFragments) {
            if (roomId == fragment.getRoom().getRoomId()) {
                return mFragments.indexOf(fragment) + 1;
            }
        }
        return -1;
    }

    @Override
    public CharSequence getPageTitle(int position) {
        Fragment fragment = getItem(position);
        if (fragment instanceof ChatListFragment) {
            return "所有聊天";
        } else if (fragment instanceof ChatFragment) {
            return ((ChatFragment) fragment).getRoom().getLastUser().getUsername();
        } else {
            return "你是谁? 从哪来?";
        }
    }

    public void sendMessage(int position, ChatMessage message) {
        Fragment fragment = getItem(position);
        if (fragment instanceof ChatFragment) {
            ((ChatFragment) fragment).newMessage(message);
        }
    }
}

File: app/src/main/java/org/miaowo/miaowo/adapter/ChatPagerAdapter.kt
package org.miaowo.miaowo.adapter

import android.support.v4.app.FragmentManager
import android.support.v4.app.FragmentPagerAdapter
import org.miaowo.miaowo.R
import org.miaowo.miaowo.base.App
import org.miaowo.miaowo.bean.data.ChatMessage
import org.miaowo.miaowo.bean.data.ChatRoom
import org.miaowo.miaowo.fragment.chat.ChatFragment
import org.miaowo.miaowo.fragment.chat.ChatListFragment

/**
 * ViewPager 的 Adapter
 * Created by luqin on 17-4-23.
 */

class ChatPagerAdapter(fm: FragmentManager) : FragmentPagerAdapter(fm) {
    private val mFragments = mutableListOf<ChatFragment>()
    private val mListFragment = ChatListFragment.newInstance()
    
    override fun getItem(position: Int) = when(position) {
        0 -> mListFragment
        else -> mFragments[position - 1]
    }

    override fun getCount() = mFragments.size + 1

    fun addChat(room: ChatRoom): Int {
        val fragment = ChatFragment.newInstance(room)
        mFragments.add(fragment)
        return mFragments.indexOf(fragment) + 1
    }

    fun getPosition(roomId: Int) = mFragments.firstOrNull { roomId == it.room?.roomId }
            ?.let { mFragments.indexOf(it) + 1 } ?: -1

    override fun getPageTitle(position: Int): CharSequence {
        val fragment = getItem(position)
        when(fragment) {
            is ChatListFragment -> return App.i.getString(R.string.chat_all)
            is ChatFragment -> return fragment.room?.lastUser?.username ?: App.i.getString(R.string.err_chat_user)
            else -> return App.i.getString(R.string.err_chat_user)
        }
    }

    fun sendMessage(position: Int, message: ChatMessage) {
        val fragment = getItem(position)
        if (fragment is ChatFragment) fragment.newMessage(message)
    }
}
